### DISCUSSION TITLE
Select existing ECR image for deploy

### DISCUSSION THREAD

**Discussion by VaZark** [aws/copilot-cli#4320] [Q&A]
I'm currently using python 3.11 as the base docker image for my projects. Since the platform is not yet supported on AWS, I need to build the container to run the tests. 

I'd like to be able to use `copilot pipeline` to,
1. Build the container
2. Run the tests
3. Push this image to ECR , if tests pass
4. Use the latest image from ECR to deploy to Fargate

The autogenerated copilot pipeline seems to use the cli deploy and I've not been able to find a flag to identify and select an image to deploy. 

Am I missing anything ? Is this feasible?

**Comment by paragbhingre** [comment]
@VaZark, I have a few questions to help me better understand your scenario.
> Since the platform is not yet supported on AWS

Which platform are you referring to here?

> The autogenerated copilot pipeline seems to use the cli deploy

What does `cli deploy` means here?

Copilot pipelines use CodeBuild for the build stage, and CloudFormation for the deploy stage.

**Comment by paragbhingre** [comment]
@VaZark  May I know the entire use case that you are trying to build here so that we can assist you better? 
The flow of your project would help us understand the problem better.

**Comment by VaZark** [comment]
Hi @paragbhingre , sorry about the late reply. 

I have a standard django backend running on python 3.10 hosted on CodeCommit. I went with docker and Fargate as python 3.10 is not supported on AWS beanstalk yet. 

The flow would be,  

1. Build the container with  the latest commit from codecommit
2. Run tests with `python manage.py test`
4. Only on success it shoud push the image to ECR and deploy to instance
5. Predeploy commands like db migrations (and rollback on failure) with `python manage.py migrate` and update static files with `python manage.py collectstatic`

**Reply by efekarakus** [reply]
Awesome, thanks @VaZark  for the detailed info!

1. Today, this is supported with the service pipeline that gets generated with `copilot pipeline init`.
2. This is also supported by modifying the generated `buildspec.yaml` file.
   For example, you can update:
   https://github.com/aws/copilot-cli/blob/dfd3767efa9ec66df24e401e190244fc0d0eb7e2/internal/pkg/template/templates/cicd/buildspec.yml#L15-L18
   With `python manage.py test`.
   I understand that today with CodeBuild there[ is no runtime-version available for python3.10](https://docs.aws.amazon.com/codebuild/latest/userguide/runtime-versions.html).
   Then, my recommendation is to modify the `install` phase to install `python3.10` within the `commands` section ([this blog](https://computingforgeeks.com/how-to-install-python-on-ubuntu-linux-system/) might help):
   https://github.com/aws/copilot-cli/blob/dfd3767efa9ec66df24e401e190244fc0d0eb7e2/internal/pkg/template/templates/cicd/buildspec.yml#L4-L14

3. Building the image is automatically supported by the Copilot generated buildspec. However, deployments will occur in separate stage in the pipeline.
4. I think this is a feature request that's the same as https://github.com/aws/copilot-cli/issues/3007. Would you mind giving a 👍 to that request? For now a workaround is to leverage `sidecars` to do database migrations before the main container is deployed (see https://github.com/aws/copilot-cli/discussions/3628#discussioncomment-3020082)

**Reply by VaZark** [reply]
Thanks for the rapid response.

I've a got a few questions/remarks based off your answer.

1. Do we need to redeploy the entire task ?
3. Can we have an option ti redeploy only the service with the dockerfile?

if we can deploy only the service, we do not need deal the ruby dependency and we can build the docker file to pushed to ECR  on codebuild

**Reply by efekarakus** [reply]
> Do we need to redeploy the entire task ?

Can you elaborate what you mean by _task_? The pipeline gets triggered for each push to the CodeCommit repository.
So anything changed in the repository (manifest files, code changes) would trigger a deployment.

> Can we have an option ti redeploy only the service with the dockerfile?

Just to clarify, do you have multiple services in your CodeCommit repository and you'd like to limit the pipeline to only a single service?
If that's the question, then you can modify [the pipeline manifest](https://aws.github.io/copilot-cli/docs/manifest/pipeline/#stages-deployments) like this:
```
stages:
  - name: test
    deployments:
       <service name>:
```
This will limit the deployment to only a single service with the given `<service name>`.

**Reply by VaZark** [reply]
Just apologising in advance as I may mix names up as I am new to ECS (or just doing things wrong :grimacing:).

I just wanted to know if I can deploy a different dockerfile without generating a new revision (of the service). 

IIRC, using the command `aws ecs update-service --cluster ${{ secrets.ECS_CLUSTER }} --service ${{ secrets.ECS_SERVICE }} --force-new-deployment` deploys the dockerfile without a new revision

**Reply by efekarakus** [reply]
> Just apologising in advance as I may mix names up as I am new to ECS (or just doing things wrong 😬).

No worries at all 😂 I totally understand that it takes a while to wrap our heads around these new concepts.

> IIRC, using the command aws ecs update-service --cluster ${{ secrets.ECS_CLUSTER }} --service ${{ secrets.ECS_SERVICE }} --force-new-deployment deploys the dockerfile without a new revision

Ah I see, yes! this is possible today with copilot. You can use the [`--force` flag with `svc deploy`](https://aws.github.io/copilot-cli/docs/commands/svc-deploy/) for example: `copilot svc deploy --force`.
There is no way of achieving this functionality within a pipeline unfortunately.

**Reply by VaZark** [reply]
Too bad. I'm usually hesitant to use `--force` as that can imply deployment without rollback in case of errors.

**Reply by efekarakus** [reply]
If I may ask, why would you need `--force` through a CI/CD pipeline? As far as I know, `--force` is mainly useful if the image is tagged with `latest` and you want to replace all the existing containers with the new image.

Copilot's pipeline never uses the `latest` tag instead each image in the generated CloudFormation template references the image digest and hence running `--force` wouldn't help in the first place. Does that make sense?
The ECS service will get updated if there are any service definition changes (such as adding a load balancer) or when a new task revision gets created (updated container image, task size, ..)
